ADR/0018-apps-continuous-perf-testing.md

```diff
@@ -0,0 +1,72 @@
+
+The approach to make this pass/fail detection possible proposed here, is to use historical results as a benchmark – not only actual results of the performance test, but also monitoring data about the resource usage, etc.
+
+> Example: Imagine you develop a web e-shop application that uses PostgreSQL data backend. Your perf test is browsing through a goods catalog and measures latency of pages. When you want to decide on pass or fail result of this test, you need to check that metrics like below are aligned to previous results of a same test with same configuration:
+>
+> - Average page load latency.
+> - Backend service CPU and memory usage...
+> - PostgreSQL CPU and memory usage…
+> - Internal PostgreSQL metrics like number of commits during the test or average number of database sessions during the test…
+>
+> If any of these metrics does not align to historical results, we mark this test result as a failure.
+
+And even if some test fails, the application team should be able to review the data and change the test definition, so next time a new test result is being evaluated based on the historical results, this new result will be one of them.
+
+The algorithm that says if the current test passes or fails when compared to historical results can vary. It can be:
+
+- Checking minimum and maximum of the metric in historical data and making sure the new result is in this range.
+- Using a floating average of historical data and setting a safe range around it and making sure the new result is in that range.
+- Some machine learning thingy.
+
+Goal of this ADR is to propose a way for this kind of testing to be implemented in Stonesoup (feature [STONE-679](https://issues.redhat.com/browse/STONE-679)). Even if it would not deprecate full-blown performance and scale testing, having a release gate with some short and small scale performance test is desirable for many application teams.
+
+We are considering two different options here. They differ in what portion of logic mentioned above is executed where.
+
+### Glossary
+
+- SUT: Subject Under Test
+- Horreum: a results storage service: <https://github.com/Hyperfoil/Horreum>
+- CPT: Continuous Performance Testing
+
+## Decision
+
+Let's use this architecture with a single Horreum instance that would be managed by Red Hat and used by all customers.
+
+![Architecture diagram with Horreum](assets/0018-apps-continuous-perf-testing.svg "Architecture diagram with Horreum")
+
+1. Performance test runs in the Tekton pipeline and generates JSON with test parameters and results.
+2. We gather configured monitoring metrics and add them to the JSON.
+3. We upload the JSON with all the data to Horreum.
+4. Horreum performs result analysis, looking for changes in configured metrics.
+4. We get PASS/FAIL decision from Horreum back to the pipeline, so pipeline can rerurn proper result.
```

Comment:
How long does it take for the analysis to be run? How is the pipeline run? Is it run as an `IntegrationTestScenario`?

Reply:
> How long does it take for the analysis to be run?

Analysis on Horreum would be (almost) instant.

> How is the pipeline run?

Actual test run would take time as per the test case. Every application have to have it's own same as every application have to have it's own functional tests.

> Is it run as an IntegrationTestScenario?

I'm not too knowledgeable about StoneSoup internals, but I assume this all will be just another Tekton PipelineRun.

Reply:
In Stonesoup, if it's an additional PipelineRun, we'd have to surface the ability to run a performance evaluation as a new option in the UI.

If it's just an additional Task within the standard PipelineRun that users are already using, then it wouldn't necessarily require additional UI/UX work. 

However, we do have "time to completion" targets for our standard PipelineRuns, so if the performance test takes 10 minutes or more, it would probably not be feasible to include it in the standard PipelineRun.

I also think these tests will be resource hogs and therefore cost us additional money (for now while the service is free) and later, cost the users additional money (in subscription fees).  We are starting out by charging users for PipelineRun minutes.  I wouldn't be surprised if a performance test costs as much as the rest of the build process in terms of PipelineRun minutes, and therefore if it was run on every build, the cost of Stonesoup could double.

That extra cost argues for making this an optional feature that our customers don't get by default.

Because of those considerations, this feature should come in through the Stonesoup feature request process rather than a technical-only ADR. Is there a prioritized feature on the STONE board for this yet?

Reply:
In addition to the cost of the PipelineRun, there would be some additional cost of activities within Horreum (I assume) as I was trying to get at in another comment: https://github.com/redhat-appstudio/book/pull/66/files/b6b5d6a4af4b7d7eb583c11bf576e34de1ae1a4f#r1133956096

Reply:
I hope I answered concerns about perf testing pipelines cost in https://github.com/redhat-appstudio/book/pull/66#issuecomment-1472303775 - TL;DR yes, there will be some - similar as for functional tests.

Reply:
Also pinged another Horreum engineer in the other thread to get Horreum resource consumption estimate.